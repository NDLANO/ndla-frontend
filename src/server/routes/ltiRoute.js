/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import defined from 'defined';
import { Helmet } from 'react-helmet';
import { BAD_REQUEST, OK } from 'http-status';
import { getHtmlLang, getLocaleObject } from '../../i18n';
import { renderPage, renderHtml } from '../helpers/render';

const bodyFields = {
  lti_message_type: {
    required: true,
    value: [
      'basic-lti-launch-request',
      'ToolProxyRegistrationRequest',
      'ContentItemSelectionRequest',
    ],
  },
  lti_version: { required: true, value: ['LTI-1p0', 'LTI-2p0'] },
  launch_presentation_return_url: { required: false },
  launch_presentation_document_target: { required: false },
  launch_presentation_height: { required: false },
  launch_presentation_width: { required: false },
};

const assets =
  process.env.NODE_ENV !== 'unittest'
    ? require(process.env.RAZZLE_ASSETS_MANIFEST) //eslint-disable-line
    : {
        client: { css: 'mock.css' },
        lti: { js: 'mock.js' },
        polyfill: { js: 'polyfill.js' },
        mathJaxConfig: { js: 'mock.js' },
      };

if (process.env.NODE_ENV === 'unittest') {
  Helmet.canUseDOM = false;
}

const getAssets = () => ({
  css: assets.client.css,
  js: [{ src: assets.lti.js }],
  polyfill: { src: assets.polyfill.js },
  mathJaxConfig: { js: assets.mathJaxConfig.js },
});

function doRenderPage(initialProps) {
  const Page = '';
  const { html, ...docProps } = renderPage(Page, getAssets(), {
    initialProps,
  });
  return { html, docProps };
}

export function parseAndValidateParameters(body) {
  let validBody = true;
  const errorMessages = [];
  Object.keys(bodyFields).forEach(key => {
    const bodyValue = body[key];
    if (bodyFields[key].required && !bodyValue) {
      validBody = false;
      errorMessages.push({ field: key, message: 'Missing required field' });
      return;
    }
    if (bodyFields[key].value && !bodyFields[key].value.includes(bodyValue)) {
      errorMessages.push({
        field: key,
        message: `Value should be one of ${bodyFields[key].value}`,
      });
      validBody = false;
    }
  });
  return validBody
    ? {
        valid: true,
        ltiData: {
          ...body,
        },
      }
    : { valid: false, messages: errorMessages };
}

export function ltiRoute(req) {
  const isPostRequest = req.method === 'POST';
  const validParameters = isPostRequest
    ? parseAndValidateParameters(req.body)
    : {};
  if (isPostRequest) {
    if (!validParameters.valid) {
      return {
        status: BAD_REQUEST,
        data: `Bad request. ${validParameters.messages
          .map(
            message => `Field ${message.field} with error: ${message.message}.`,
          )
          .join(', ')}`,
      };
    }
  }

  const lang = getHtmlLang(defined(req.params.lang, ''));
  const locale = getLocaleObject(lang);

  const { html, docProps } = doRenderPage({
    locale,
    ltiData: validParameters.ltiData,
  });

  return renderHtml(req, html, { status: OK }, docProps);
}
