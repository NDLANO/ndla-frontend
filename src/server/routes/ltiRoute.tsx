/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { Request } from "express";
import { FilledContext, HelmetProvider } from "react-helmet-async";
import { getHtmlLang, getLocaleObject } from "../../i18n";
import { BAD_REQUEST, OK } from "../../statusCodes";
import { Assets } from "../helpers/Document";
import { renderPage, renderHtml } from "../helpers/render";

const bodyFields: Record<string, { required: boolean; value?: any }> = {
  lti_message_type: {
    required: true,
    value: ["basic-lti-launch-request", "ToolProxyRegistrationRequest", "ContentItemSelectionRequest"],
  },
  lti_version: { required: true, value: ["LTI-1p0", "LTI-2p0"] },
  launch_presentation_return_url: { required: false },
  launch_presentation_document_target: { required: false },
  launch_presentation_height: { required: false },
  launch_presentation_width: { required: false },
};

//@ts-ignore
const assets = require(process.env.ASSETS_MANIFEST);

if (process.env.NODE_ENV === "unittest") {
  HelmetProvider.canUseDOM = false;
}

const getAssets = (): Assets => ({
  css: assets["client.css"],
  js: [{ src: assets["lti.js"]! }],
  mathJaxConfig: { js: assets["mathJaxConfig.js"]! },
});

function doRenderPage<T extends object>(initialProps: T) {
  //@ts-ignore
  const helmetContext: FilledContext = {};
  const Page = <HelmetProvider context={helmetContext}>{""}</HelmetProvider>;
  const { html, ...docProps } = renderPage(Page, getAssets(), {
    initialProps,
  });
  return { html, docProps, helmetContext };
}

export function parseAndValidateParameters(body: any) {
  let validBody = true;
  const errorMessages: { field: string; message: string }[] = [];
  Object.keys(bodyFields).forEach((key) => {
    const bodyValue = body[key];
    if (bodyFields[key]?.required && !bodyValue) {
      validBody = false;
      errorMessages.push({ field: key, message: "Missing required field" });
      return;
    }
    if (bodyFields[key]?.value && !bodyFields[key]?.value.includes(bodyValue)) {
      errorMessages.push({
        field: key,
        message: `Value should be one of ${bodyFields[key]?.value}`,
      });
      validBody = false;
    }
  });
  return validBody
    ? {
        valid: true,
        ltiData: {
          ...body,
        },
      }
    : { valid: false, messages: errorMessages };
}

export function ltiRoute(req: Request) {
  const isPostRequest = req.method === "POST";
  const validParameters = isPostRequest ? parseAndValidateParameters(req.body) : undefined;
  if (isPostRequest) {
    if (!validParameters?.valid) {
      const messages = validParameters?.messages
        ?.map((msg) => `Field ${msg.field} with error: ${msg.message}.`)
        .join(",");
      return {
        status: BAD_REQUEST,
        data: `Bad request. ${messages}`,
      };
    }
  }

  const lang = getHtmlLang(req.params.lang ?? "");
  const locale = getLocaleObject(lang).abbreviation;

  const { html, docProps, helmetContext } = doRenderPage({
    locale,
    ltiData: validParameters?.ltiData,
  });

  return renderHtml(html, { status: OK }, docProps, helmetContext);
}
