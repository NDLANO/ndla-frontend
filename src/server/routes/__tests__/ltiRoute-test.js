/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ltiRoute, parseAndValidateParameters } from "../ltiRoute";

test("ltiRoute 200 OK ", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRoute({
    params: {
      lang: "nb",
      articleId: "26050",
      resourceId: "urn:resource:123",
    },
    body,
    method: "POST",
    headers: {
      "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
    },
  });

  expect(response.status).toBe(200);
});

test("ltiRoute 200 OK only required params", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
  };
  const response = await ltiRoute({
    params: {
      lang: "nb",
      articleId: "26050",
      resourceId: "urn:resource:123",
    },
    body,
    method: "POST",
    headers: {
      "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
    },
  });

  expect(response.status).toBe(200);
});

test("ltiRoute 400 BAD REQUEST", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRoute({
    params: {
      lang: "nb",
      articleId: "26050",
      resourceId: "urn:resource:123",
    },
    method: "POST",
    body,
    headers: {
      "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
    },
  });

  expect(response).toMatchSnapshot();
});

test("ltiRoute 400 BAD REQUEST wrong values", async () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "wrong version",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const response = await ltiRoute({
    params: {
      lang: "nb",
      articleId: "26050",
      resourceId: "urn:resource:123",
    },
    method: "POST",
    body,
    headers: {
      "user-agent": "Mozilla/5.0 Gecko/20100101 Firefox/58.0",
    },
  });

  expect(response).toMatchSnapshot();
});

test("parseAndValidateParameters no errors", () => {
  const body = {
    lti_message_type: "basic-lti-launch-request",
    lti_version: "LTI-1p0",
    launch_presentation_return_url: "http://ndla-api/some-return-url",
    launch_presentation_document_target: "iframe",
    launch_presentation_height: "800",
    launch_presentation_width: "1200",
  };
  const result = parseAndValidateParameters(body);
  expect(result).toMatchSnapshot();
});

test("parseAndValidateParameters errors", () => {
  const body = {};
  const result = parseAndValidateParameters(body);
  expect(result).toMatchSnapshot();
});
