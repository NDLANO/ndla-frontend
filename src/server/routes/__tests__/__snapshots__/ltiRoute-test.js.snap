// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ltiRoute 400 BAD REQUEST 1`] = `
Object {
  "data": "Bad request. Field lti_version with error: Missing required field.",
  "status": 400,
}
`;

exports[`ltiRoute 400 BAD REQUEST wrong values 1`] = `
Object {
  "data": "Bad request. Field lti_version with error: Value should be one of LTI-1p0,LTI-2p0.",
  "status": 400,
}
`;

exports[`parseAndValidateParameters errors 1`] = `
Object {
  "messages": Array [
    Object {
      "field": "lti_message_type",
      "message": "Missing required field",
    },
    Object {
      "field": "lti_version",
      "message": "Missing required field",
    },
  ],
  "valid": false,
}
`;

exports[`parseAndValidateParameters no errors 1`] = `
Object {
  "ltiData": Object {
    "launch_presentation_document_target": "iframe",
    "launch_presentation_height": "800",
    "launch_presentation_return_url": "http://ndla-api/some-return-url",
    "launch_presentation_width": "1200",
    "lti_message_type": "basic-lti-launch-request",
    "lti_version": "LTI-1p0",
  },
  "valid": true,
}
`;
